sys/net/gnrc/conn/tcp/gnrc_conn_tcp.c

```diff
@@ -0,0 +1,560 @@
+
+    /* Guard: Return is already closed */
+    if(conn->state == CLOSED){
+        return 0;
+    }
+
+    /* Guard: Caller must be the connections owner */
+    if (conn->owner != thread_getpid()) {
+        return -EPERM;
+    }
+
+    /* Guard: Only a opened connection can be closed */
+    if (conn->state != LISTEN && conn->state != SYN_RCVD
+        && conn->state != ESTABLISHED && conn->state != CLOSE_WAIT) {
+        return -EALREADY;
+    }
+
+    /* If the connection is passive and the handle is not accepted */
+    if (conn->status & GNRC_TCP_STATUS_PASSIVE && !(conn->status & GNRC_TCP_STATUS_ACCEPTED)) {
+        return -EBADF;
+    }
+
+    /* Purge message queue */
+    while(msg_try_receive(&msg) >= 0){
+    }
+
+    /* Setup Userspace Timeout */
+    msg_tout.type = MSG_TYPE_USER_TIMEOUT;
+    xtimer_set_msg(&usr_tout, GNRC_TCP_USER_TIMEOUT_DURATION, &msg_tout, thread_getpid());
+
+    /* Start closing sequence */
+    _fsm(conn, CALL_CLOSE, NULL, NULL, 0);
+
+    /* Wait until the connection has been closed */
+    while (conn->state != CLOSED) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case MSG_TYPE_USER_TIMEOUT:
+                DEBUG("gnrc_conn_tcp : conn_tcp_close() : USER_TIMEOUT\n");
+                _fsm(conn, TIMEOUT_USER, NULL, NULL, 0);
+                err = -ETIME;
+                break;
+
+            case MSG_TYPE_NOTIFY_USER:
+                DEBUG("gnrc_conn_tcp : conn_tcp_close() : MSG_TYPE_NOTIFY_USER\n");
+                break;
+
+            default:
+                DEBUG("gnrc_conn_tcp : conn_tcp_close() : other message type\n");
+        }
+    }
+
+    /* Stop User Space Timeout if it didn't expire earlier */
+    if (err != -ETIME) {
+        xtimer_remove(&usr_tout);
+    }
+
+    /* If this was a passive connection, re-open */
+    if (conn->status & GNRC_TCP_STATUS_PASSIVE) {
+        _fsm(conn, CALL_OPEN, NULL, NULL, 0);
+    }
+    return err;
+}
+
+
+int conn_tcp_getlocaladdr(conn_tcp_t *conn, void *addr, uint16_t *port)
+{
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+    memcpy(addr, &(conn->local_addr), conn->local_addr_len);
+    *port = conn->local_port;
+    return conn->local_addr_len;
+}
+
+
+int conn_tcp_getpeeraddr(conn_tcp_t *conn, void *addr, uint16_t *port)
+{
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+    memcpy(addr, &(conn->peer_addr), conn->peer_addr_len);
+    *port = conn->peer_port;
+    return conn->peer_addr_len;
+}
+
+int conn_tcp_connect(conn_tcp_t *conn, const void *addr, size_t addr_len, uint16_t port)
+{
+    msg_t msg;
+    msg_t msg_tout;
+    xtimer_t usr_tout;
+    int err = 0;
+
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+
+    /* Connection is already connected: Return -EISCONN */
+    if (conn->state != CLOSED) {
+        return -EISCONN;
+    }
+
+    /* This call is only allowed on active Connections*/
+    if (conn->status & GNRC_TCP_STATUS_PASSIVE) {
+        return -EOPNOTSUPP;
+    }
```

Comment:
All `-EISCONN`, `-EOPNOTSUPP` could (and maybe should) be thrown from the state machine since this is essential to the state process itself.